sdk/python/feast/permissions/auth_model.py

```diff
@@ -11,9 +11,12 @@ class OidcAuthConfig(AuthConfig):
     auth_discovery_url: str
     client_id: str
     client_secret: Optional[str] = None
+    realm: str = "master"
+
+
+class OidcClientAuthConfig(OidcAuthConfig):
     username: str
     password: str
```

Comment:
After reviewing the code again, looks like there are other fields that are needed only at the client side.

Clients use these fields to get the token:
```py
        token_request_body = {
            "grant_type": "password",
            "client_id": self.auth_config.client_id,
            "client_secret": self.auth_config.client_secret,
            "username": self.auth_config.username,
            "password": self.auth_config.password,
        }
```
On the other hand, servers use these fields to validate the token:
```py
        self.oidc_discovery_service = OIDCDiscoveryService(
            self._auth_config.auth_discovery_url
        )
...
            client_id = self._auth_config.client_id
```

So, to summarize, servers config is like:
```yaml
auth:
  type: oidc
  auth_discovery_url: http://0.0.0.0:9999/realms/rbac_example/.well-known/openid-configuration
  client_id: app
```

Client config template:
```yaml
  auth_discovery_url: http://0.0.0.0:9999/realms/rbac_example/.well-known/openid-configuration
  client_id: app
  client_secret: Rw4X56WaHODmiYw8duPPTwuH0OoCtVvg
  password: password
  type: oidc
  username: reader
```

`realm` field is no more needed since it is already part of the `auth_discovery_url` (it was needed before when we built the URL programmatically).

Reply:
Thanks for all the details. I have updated the PR. PTAL